### ISSUE TITLE
ValidationError is not Hashable (2.0.0)

### ISSUE THREAD

**Issue by Guibod** [schematics/schematics#452]
I can't `logging.exception(ValidationError)` it seems that the object is not hashable.

That's sad, it used to work pretty well on 1.1. I'm using Python 3.4.

```
class TestUnhashable(unittest.TestCase):
    def test_type_error_is_hashable(self):
        try:
            raise TypeError('Foo bar')
        except TypeError as e:
            logging.exception(e)

    def test_validation_error_is_unhashable(self):
        try:
            raise ValidationError('Foo bar')
        except ValidationError as e:
            with self.assertRaises(TypeError) as c:
                logging.exception(e)

            self.assertIsInstance(c.exception, TypeError)
            self.assertEqual(c.exception.args[0], 'unhashable type: \'ValidationError\'')
```

**Comment by Guibod** [comment]
Correction. I have no clue if it was working in v1.1. Yet this would be great to be able to `logging.exception(ValidationError)`

**Comment by lkraider** [comment]
Your code works in Python 2, but breaks in Python 3 because it auto-sets `__hash__ = None` when `__eq__` is defined for a class (as is the case in Schematics `ErrorMessage`) - you need to provide your own hash method.

In Python an object can only be [hashable](https://docs.python.org/3/glossary.html#term-hashable) if it has a hash value which never changes during its lifetime, that is, it is immutable.

We have a "problem" right now with the Errors in that they are mutable (they contain a list of messages and a `pop` method that changes it).

Also, hashable objects which compare equal must have the same hash value, and we use the instance `__dict__` for comparison inside `__eq__`, which is also mutable/unhashable.

Python itself does not require Exceptions to be immutable. At least in 2002, GVR was in favor of mutable exceptions: https://mail.python.org/pipermail/python-dev/2002-August/027900.html

The `logging` framework assumes exceptions are hashable (and thus immutable), which is an undocumented assumption. But since logging is in the stdlib, we need to play by their rules.

The fix to support logging is to make all schematics exceptions immutable. I don't have a proposal for how to do this, but would welcome a PR.

**Comment by lkraider** [comment]
Fixed in #477
